File: src/main/java/software/aws/glue/tableversions/utils/DDBUtil.java
// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

package software.aws.glue.tableversions.utils;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.dynamodbv2.document.PutItemOutcome;
import com.amazonaws.services.dynamodbv2.document.Table;

/**
 * This is a utility class with methods to write items to DynamoDB table. from /
 * to a DynamoDB table.
 * 
 * @author Ravi Itha, Amazon Web Services, Inc.
 *
 */
public class DDBUtil {

	public boolean insertCleanupStatusToDynamoDB(AmazonDynamoDB ddbClient, String ddbTableName, String hashKey,
			String rangeKey, long executionId, String executionBatchId, String databaseName, String tableName,
			int numTableVersionsB4Cleanup, int numVersionsRetained, int numDeletedVersions) {

		boolean itemInserted = false;
		DynamoDB dynamoDB = new DynamoDB(ddbClient);
		Table table = dynamoDB.getTable(ddbTableName);
		Item item = new Item()
				.withPrimaryKey(hashKey, executionId)
				.withNumber(rangeKey, Long.parseLong(executionBatchId))
				.withString("table_name", tableName)
				.withString("database_name", databaseName)
				.withNumber("number_of_versions_before_cleanup", numTableVersionsB4Cleanup)
				.withNumber("number_of_versions_retained", numVersionsRetained)
				.withNumber("number_of_versions_deleted", numDeletedVersions);
		// Write the item to the table
		PutItemOutcome outcome = table.putItem(item);
		int statusCode = outcome.getPutItemResult().getSdkHttpMetadata().getHttpStatusCode();
		if (statusCode == 200) {
			itemInserted = true;
			System.out.println("Table version inserted to DynamoDB table: " + ddbTableName);
		}
		return itemInserted;
	}

	/**
	 * Method to write Table version details to a DynamoDB table.
	 * 
	 * @param dynamoDBClient
	 * @param dynamoDBTblName
	 * @param primaryPartKey
	 * @param primarySortKey
	 * @param executionBatchId
	 * @param databaseName
	 * @param tableName
	 * @param notificationTime
	 * @return
	 */
	public boolean insertTableDetailsToDynamoDB(AmazonDynamoDB ddbClient, String ddbTableName, String hashKey,
			String rangeKey, long executionBatchId, String databaseName, String tableName, String messageSentTime) {

		boolean itemInserted = false;
		DynamoDB dynamoDB = new DynamoDB(ddbClient);
		Table table = dynamoDB.getTable(ddbTableName);
		Item item = new Item().withPrimaryKey(hashKey, executionBatchId)
				.withString(rangeKey, databaseName.concat("|").concat(tableName)).withString("table_name", tableName)
				.withString("database_name", databaseName).withString("message_sent_time", messageSentTime);
		// Write the item to the table
		PutItemOutcome outcome = table.putItem(item);
		int statusCode = outcome.getPutItemResult().getSdkHttpMetadata().getHttpStatusCode();
		if (statusCode == 200) {
			itemInserted = true;
			System.out.println("Table version inserted to DynamoDB table: " + ddbTableName);
		}
		return itemInserted;
	}

}

File: src/utils/ddb_util.py
# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

import boto3
import os


def get_dynamodb_client():
    """
    DynamoDB Client
    
    :return:
    """
    return boto3.client('dynamodb')


def insert_to_planner_dynamodb(client, ddb_table_name, hash_key, execution_batch_id, range_key,
                               database_name, table_name, message_sent_time):
    """
    Inserts to Planner DynamoDB table

    :param client:
    :param ddb_table_name:
    :param hash_key:
    :param execution_batch_id:
    :param range_key:
    :param database_name:
    :param table_name:
    :param message_sent_time:
    :return:
    """
    client.put_item(
        TableName=ddb_table_name,
        Item={
            hash_key: {'N': str(execution_batch_id)},
            range_key: {'S': database_name + '|' + table_name},
            'table_name': {'S': table_name},
            'database_name': {'S': database_name},
            'message_sent_time': {'S': message_sent_time}
        })


def insert_to_cleanup_dynamodb(client, ddb_table_name, hash_key, range_key,
                               execution_id, execution_batch_id, database_name, table_name,
                               num_of_versions_before_cleanup, num_of_versions_retained, num_of_versions_deleted):
    """
    Inserts to Cleanup DynamoDB table

    :param client:
    :param ddb_table_name:
    :param hash_key:
    :param range_key:
    :param execution_id:
    :param execution_batch_id:
    :param database_name:
    :param table_name:
    :param num_of_versions_before_cleanup:
    :param num_of_versions_retained:
    :param num_of_versions_deleted:
    :return:
    """
    client.put_item(
        TableName=ddb_table_name,
        Item={
            hash_key: {'N': str(execution_id)},
            range_key: {'N': str(execution_batch_id)},
            'table_name': {'S': table_name},
            'database_name': {'S': database_name},
            'number_of_versions_before_cleanup': {'N': str(num_of_versions_before_cleanup)},
            'number_of_versions_retained': {'N': str(num_of_versions_retained)},
            'number_of_versions_deleted': {'N': str(num_of_versions_deleted)}
        })
